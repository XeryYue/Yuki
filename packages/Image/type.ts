import { ComputedRef } from 'vue'

export type BrowserColors = {
  color: string
  barBgColor: string
  inputBgColor: string
  borderColor: string
  titleColor: string
}

type HeadAttrs = {
  head: string
  show: boolean
}

export type ImageProvide = {
  setColors: ComputedRef<BrowserColors>
  showLinkType: ComputedRef<string>
  setHead: ComputedRef<HeadAttrs>
}

export const READONLY_IMAGE_KEY = 'imageKey'
