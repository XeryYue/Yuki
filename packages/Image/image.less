.fect-image {
  position: relative;
  display: inline-block;
  overflow: hidden;
  max-width: 100%;
  border-radius: var(--fay-radius);
  img {
    object-fit: scale-down;
    display: block;
    width: 100%;
    height: 100%;
  }
}

.fect-image {
  .skeleton {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    width: 100%;
    height: 100%;
    background-image: linear-gradient(
      270deg,
      var(--accents-1),
      var(--accents-2),
      var(--accents-2),
      var(--accents-1)
    );
    background-size: 400% 100%;
    animation: loading 3s ease-in-out infinite;
    transition: opacity 300ms ease-out;
  }
  @keyframes loading {
    0% {
      background-position: 200% 0;
    }
    to {
      background-position: -200% 0;
    }
  }
}

.fect-image__browser {
  background-color: transparent;
  box-shadow: var(--fay-shadowLarge);
  width: max-content;
  max-width: 100%;
  margin: 0 auto;
  border-radius: var(--fay-radius);
  overflow: hidden;
  .fect-image {
    border-top-left-radius: 0;
    border-top-right-radius: 0;
  }
  header {
    height: 40px;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    border-bottom: 1px solid var(--accents-2);

    .traffic {
      width: auto;
      position: absolute;
      left: var(--fay-gap-half);
      top: 50%;
      transform: translateY(-50%);
      bottom: 0;
      height: 100%;
      display: flex;
      align-items: center;
      user-select: none;
      span {
        border-radius: 50%;
        width: 12px;
        height: 12px;
        display: inline-block;
        margin-right: 8px;
      }
      .close {
        background-color: #ff5f56;
      }

      .mini {
        background-color: #ffbd2e;
      }

      .full {
        background-color: #27c93f;
      }
    }

    .title {
      font-size: 12px;
    }
  }
}

.fect-image__browser {
  .address-input {
    height: 28px;
    min-width: 40%;
    max-width: 60%;
    color: inherit;
    border-radius: 3px;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0 10px;
    overflow: hidden;
    position: relative;
    font-size: 12px;
    color: inherit;
    a {
      max-width: 90%;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
      display: inline-block;
      color: inherit;
    }
    .https {
      width: 12px;
      height: 12px;
      margin-right: 5px;
      user-select: none;
      margin-top: -1px;
      color: inherit;
    }
  }
}
