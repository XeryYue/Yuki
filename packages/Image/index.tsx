import { computed, ref, watchEffect, CSSProperties } from 'vue'
import { createNameSpace } from '../utils'
import ImgSkeleton from './image-skeleton'
import './image.less'

const [createComponent] = createNameSpace('Image')

export default createComponent({
  props: {
    src: {
      type: String,
      default: '',
    },
    skeleton: Boolean,
    maxDelay: {
      type: [String, Number],
      default: 3000,
    },
    width: {
      type: [Number, String],
      default: 'auto',
    },
    height: {
      type: [Number, String],
      default: 'auto',
    },
  },
  setup(props, { attrs }) {
    const imgRef = ref<HTMLImageElement>()
    const loading = ref<boolean>(true)
    const showSkeleton = ref(props.skeleton)

    watchEffect((onInvalidate) => {
      const timer = setTimeout(() => {
        if (!imgRef.value) return
        if (imgRef.value.complete) {
          loading.value = false
          showSkeleton.value = false
        }
      }, Number(props.maxDelay))
      onInvalidate(() => {
        clearTimeout(timer)
      })
    })

    const setStyle = computed(() => {
      const { width, height } = props
      const style: CSSProperties = {
        width,
        height,
      }
      return style
    })

    return () => (
      <div class="fect-image" style={setStyle.value}>
        {showSkeleton.value && (
          <ImgSkeleton opacity={loading.value ? 0.5 : 0} />
        )}
        <img
          ref={imgRef}
          src={props.src}
          width={props.width}
          height={props.height}
          style={{ display: !showSkeleton.value ? 'block' : 'none' }}
          {...attrs}
        />
      </div>
    )
  },
})
