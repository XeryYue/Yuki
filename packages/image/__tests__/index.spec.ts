import { mount } from '@vue/test-utils'
import Image from '../index'

describe('Image', () => {
  it('should be render as element', () => {
    const img = mount(Image)
    expect(img.html()).toMatchSnapshot()
    expect(() => img.unmount()).not.toThrow()
  })
  it('should  support skeleton', () => {
    const wrapper = mount(Image, {
      props: {
        skeleton: true,
      },
    })
    expect(wrapper.find('.skeleton')).toBeTruthy()
  })
  it('should support useBrowser', () => {
    const wrapper = mount(Image, {
      props: {
        useBrowser: true,
        height: '240px',
        width: '540px',
      },
    })
    expect(wrapper.find('.fect-image__browser')).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
