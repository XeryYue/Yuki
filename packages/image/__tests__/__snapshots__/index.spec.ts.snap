// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should be render as element 1`] = `
<div class="fect-image" style="width: auto; height: auto;">
  <!----><img src="" width="0" height="0" style="display: block;">
</div>
`;

exports[`Image should support useBrowser 1`] = `
<div class="fect-image__browser">
  <header>
    <div class="traffic"><span class="close" role="browser-close"></span><span class="mini" role="browser-mini"></span><span class="full" role="browser-full"></span></div>
    <div class="title"><span></span></div>
  </header>
  <div class="fect-image" style="width: 540px; height: 240px;">
    <!----><img src="" width="0" height="0" style="display: block;">
  </div>
</div>
`;
